packages/xo-server/src/xapi/mixins/patching.js

```diff
@@ -464,4 +465,97 @@ export default {
     // host-by-host install
     throw new Error('non pool-wide install not implemented')
   },
+
+  async rollingPoolUpdate() {
+    const hosts = filter(this.objects.all, { $type: 'host' })
+    await Promise.all(hosts.map(host => host.$call('assert_can_evacuate')))
+
+    log.debug('Install patches')
+    await this.installPatches()
+
+    // Remember on which hosts the running VMs are
+    const vmsByHost = mapValues(
+      groupBy(
+        filter(this.objects.all, {
+          $type: 'VM',
+          power_state: 'Running',
+          is_control_domain: false,
+        }),
+        vm => {
+          const hostId = vm.$resident_on?.$id
+
+          if (hostId === undefined) {
+            throw new Error('Could not find host of all running VMs')
+          }
+
+          return hostId
+        }
+      ),
+      vms => vms.map(vm => vm.$id)
+    )
+
+    // Put master in first position to restart it first
+    const indexOfMaster = hosts.findIndex(host => host.$ref === this.pool.master)
+    if (indexOfMaster > 0) {
+      ;[hosts[0], hosts[indexOfMaster]] = [hosts[indexOfMaster], hosts[0]]
+    }
+
+    // Restart all the hosts one by one
+    for (const host of hosts) {
+      const hostId = host.uuid
+      const metricsRef = host.metrics
+
+      await this.barrier(metricsRef)
+      await this._waitObjectState(metricsRef, metrics => metrics.live)
+      const rebootTime = parseDateTime(await this.call('host.get_servertime', host.$ref))
+
+      log.debug(`Evacuate and restart host ${hostId}`)
+      await this.rebootHost(hostId)
+
+      log.debug(`Wait for host ${hostId} to be up`)
+      await timeout.call(
+        (async () => {
+          await this._waitObjectState(
+            hostId,
+            host => host.enabled && rebootTime < host.other_config.agent_start_time * 1e3
+          )
+          await this._waitObjectState(metricsRef, metrics => metrics.live)
+        })(),
+        this._restartHostTimeout,
+        new Error(`Host ${hostId} took too long to restart`)
+      )
+      log.debug(`Host ${hostId} is up`)
+    }
+
+    log.debug('Migrate VMs back to where they were')
+
+    // Start with the last host since it's the emptiest one after the rolling
+    // update
+    ;[hosts[0], hosts[hosts.length - 1]] = [hosts[hosts.length - 1], hosts[0]]
```

Comment:
Why not simply reverse?

Reply:
For better code readability? Otherwise I don't need to reverse the whole array, that's why. I only need the last host to be in first position since it doesn't have any running VMs on it.

Reply:
Ok :slightly_smiling_face: